### PR TITLE
Enable automatic limit for array-as-uninterpreted-function [depends-on: #1874]

### LINKED ISSUE #2018: bv_cbmct::unbounded_arrayt::U_AUTO is never enabled
While seemingly meant to be the default (only never/always can be set via the command line), the default is actually `U_NONE`. Enabling it is trivial, but will break lots of examples until #1086 is complete and merged. Furthermore it seems current code only looks at the top-level size, not the nested values. Below is a patch that may be a useful starting point:
````c++
diff --git a/src/cbmc/cbmc_solvers.cpp b/src/cbmc/cbmc_solvers.cpp
index ab69f9b..caf40be 100644
--- a/src/cbmc/cbmc_solvers.cpp
+++ b/src/cbmc/cbmc_solvers.cpp
@@ -110,6 +110,8 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_default()
     bv_cbmc->unbounded_array=bv_cbmct::unbounded_arrayt::U_NONE;
   else if(options.get_option("arrays-uf")=="always")
     bv_cbmc->unbounded_array=bv_cbmct::unbounded_arrayt::U_ALL;
+  else
+    bv_cbmc->unbounded_array=bv_cbmct::unbounded_arrayt::U_AUTO;

   solver->set_prop_conv(std::move(bv_cbmc));

diff --git a/src/solvers/flattening/boolbv.cpp b/src/solvers/flattening/boolbv.cpp
index 4cf43a0..503be96 100644
--- a/src/solvers/flattening/boolbv.cpp
+++ b/src/solvers/flattening/boolbv.cpp
@@ -689,17 +689,20 @@ bool boolbvt::is_unbounded_array(const typet &type) const
   if(unbounded_array==unbounded_arrayt::U_ALL)
     return true;

+#if 0
   const exprt &size=to_array_type(type).size();

   mp_integer s;
   if(to_integer(size, s))
     return true;
+#endif

+  bool retval=false;
   if(unbounded_array==unbounded_arrayt::U_AUTO)
-    if(s>MAX_FLATTENED_ARRAY_SIZE)
-      return true;
+    if(boolbv_width(type)>MAX_FLATTENED_ARRAY_SIZE)
+      retval=true;

-  return false;
+  return retval;
 }

 void boolbvt::print_assignment(std::ostream &out) const
````

### PR DESCRIPTION
**Pull request by tautschnig**
Previously, the command line permitted setting uninterpreted functions
to "never" or "always", where "never" actually was the default. The
"automatic" mode could not be enabled in any way.

This reverts #6232, but now includes additional changes to the array
theory to handle nested struct members no different from members of a
top-level struct.

Fixes: #2018

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8089c08132bc07cd24f98cd03da7defbf327a3ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108222857

**Comment by tautschnig** [comment]
Reviews would be appreciated, but still marking do-not-merge as benchmarking is required. (Initial benchmarking on ReachSafety-Arrays suggests that, as expected, the memory footprint grows.)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 35d2f2c31613448428917956ca870b1287382580).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113713932

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4615ed1ac89a8c2a28277fc8b1f934c67f6db3d4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113976199

**Comment by tautschnig** [comment]
I have now run this on SV-COMP, and did notice a speedup in some cases, but also a single-digit number of benchmarks newly running out of memory. That's not unexpected, but it does suggest we should put #1874 in place before merging this one.

**Comment by TGWDB** [comment]
As discussed in #6130 it would probably be good to merge this PR soon. With that in mind, I've locally rebuilt, rebase, and run unit/regression tests and not seen any problems (or major performance changes). If there is no blocker remaining perhaps this can be rebased, pushed and reviewed for a merge to resole #6130.

**Comment by TGWDB** [comment]
> As discussed in #6130 it would probably be good to merge this PR soon. With that in mind, I've locally rebuilt, rebase, and run unit/regression tests and not seen any problems (or major performance changes). If there is no blocker remaining perhaps this can be rebased, pushed and reviewed for a merge to resole #6130.

@tautschnig Any issues with merging this as is?

@kroening @peterschrammel @martin-cs Codeowner approval/disapproval?

File: src/solvers/flattening/boolbv_width.cpp
Comment (martin-cs) [inline src/solvers/flattening/boolbv_width.cpp]: I feel l;ke maybe this would like to become an invariant?

**Review by martin-cs — APPROVED** [review]
I am OK with this but there is some kind of bug in equality of unbounded arrays which I have yet to reliably trigger via C or locate.  I figure that this might expose it so it would be good to have a few tests of what happens when an array is over `MAX_FLATTENED_ARRAY_SIZE`

**Comment by TGWDB** [comment]
Closing as I believe this has been covered by #6194 and fix in #6232. Please reopen if you believe this is erroneous.

**Comment by tautschnig** [comment]
Note to self: `cbmc regression/cbmc/bounds_check1/main.c --bounds-check --pointer-check` is one of the examples now taking a lot longer, though really it is SSA conversion that is consuming a lot of time. To be debugged.

**Comment by TGWDB** [comment]
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

**Pull request closed without merging**
